File: BinaryExpressionTree/python/binary_expression_tree.py
import typing
import traceback

E = typing.TypeVar("E")
Operator = typing.Literal["+", "-", "/", "*"]
Operators = typing.get_args(Operator)


class Node(typing.Generic[E]):

    def __init__(
            self,
            element: E,
            parent: typing.Optional["Node[E]"] = None,
            left: typing.Optional["Node[E]"] = None,
            right: typing.Optional["Node[E]"] = None,
    ) -> None:
        self._element = element
        self._parent = parent
        self._left = left
        self._right = right
        self._index = -1  # only computed when generating tree visualization

    @property
    def element(self) -> E:
        return self._element

    @property
    def index(self) -> int:
        return self._index

    @index.setter
    def index(self, i: int) -> None:
        self._index = i

    @property
    def parent(self) -> typing.Union["Node[E]", None]:
        return self._parent

    @parent.setter
    def parent(self, node: "Node[E]") -> None:
        if self._parent:
            raise ValueError("Parent is not empty!")
        self._parent = node

    @property
    def left(self) -> typing.Union["Node[E]", None]:
        return self._left

    @left.setter
    def left(self, node: "Node[E]") -> None:
        if self._left:
            raise ValueError("Left is not empty!")
        self._left = node

    @property
    def right(self) -> typing.Union["Node[E]", None]:
        return self._right

    @right.setter
    def right(self, node: "Node[E]") -> None:
        if self._right:
            raise ValueError("Right is not empty!")
        if self._left is None:
            raise ValueError("Left is empty!")
        self._right = node

    def children(self) -> list["Node[E]"]:
        output = []
        if self._left:
            output.append(self._left)
            if self._right:
                output.append(self._right)
        return output

    def has_children(self) -> bool:
        return len(self.children()) > 0

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}({self._element})"


class BinaryExpressionTree:

    def __init__(self, expression: str) -> None:
        self._expression = expression.replace(" ", "")
        self._index = 0
        self._root: typing.Union[Node[str], None] = None
        self._validate()

    @property
    def depth(self) -> int:
        return len(self.get_layers(self._root))

    @property
    def height(self) -> int:
        layers = self.get_layers(self._root)
        return sum(len(lst) for lst in layers)

    @property
    def size(self) -> int:
        return self.height

    @property
    def empty(self) -> bool:
        return self._root is None

    @property
    def root(self) -> Node[str]:
        return self._root

    def inorder_traversal(self, node: typing.Optional[Node[E]] = None) -> list[Node[E]]:
        if self.empty:
            return []
        if node is None:
            node = self._root
        output = []
        if node.left:
            output.extend(self.postorder_traversal(node.left))
        output.append(node)
        if node.right:
            output.extend(self.postorder_traversal(node.right))
        return output

    def preorder_traversal(self, node: typing.Optional[Node[E]] = None) -> list[Node[E]]:
        if self.empty:
            return []
        if node is None:
            node = self._root
        output = [node]

        children = node.children()
        for child in children:
            output.extend(self.preorder_traversal(child))

        return output

    def postorder_traversal(self, node: typing.Optional[Node[E]] = None) -> list[Node[E]]:
        if self.empty:
            return []
        if node is None:
            node = self._root
        if self.is_leaf(node):
            return [node]
        output = []
        if node.left:
            output.extend(self.postorder_traversal(node.left))
        if node.right:
            output.extend(self.postorder_traversal(node.right))
        output.append(node)
        return output

    def breath_first_traversal(self) -> list[Node[E]]:
        layers = self.get_layers(self._root)
        return [node for lst in layers for node in lst]

    @staticmethod
    def visualize_left_to_right(node: typing.Optional[Node[E]], level: typing.Optional[int]) -> str:

        def _vis(n: Node[E], lvl: int) -> list[str]:
            output = []
            if n.right:
                output.extend(_vis(n.right, lvl + 2))
            output.append(f"""{((" " * 4) * lvl)}-> {n.element}""")
            if n.left:
                output.extend(_vis(n.left, lvl + 2))
            return output

        if node is None:
            return ""
        results = _vis(node, level)
        results.insert(0, "\n")
        results.append("\n")
        return "\n".join(results)

    @staticmethod
    def visualize_top_to_bottom(node: typing.Optional[Node[E]], depth: int, start: typing.Optional[int] = 0) -> str:

        def _create_index(input_: list[list[Node[E]]]) -> int:
            flattened = [n for lst in input_ for n in lst]
            for idx, n in enumerate(flattened):
                n.index = idx
            return len(flattened)

        def _vis(n: Node[E], d: int, s: int) -> None:
            val = f"{n.element}" if n is not None else " "
            col = d ** 2
            n_str = f"""{" " * col}{" " * s}{val}"""
            results[n.index] = n_str
            if n.left:
                _vis(n.left, d - 1, s)
                if n.right:
                    _vis(n.right, d - 1, col + s)

        layers = BinaryExpressionTree.get_layers(node)
        results = [""] * _create_index(layers)
        _vis(node, depth, start)
        output = [""]
        for i in range(depth):
            n = len(layers[i])
            lst = []
            for _ in range(n):
                try:
                    val = results.pop(0)
                    split = val.split(" ")
                    div = len("".join(lst))
                    lst.append(" ".join(split[div:]))
                except IndexError:
                    break
            output.append("".join(lst))
        output.append("\n")
        return "\n\n".join(output)

    @staticmethod
    def is_leaf(node: Node[E]) -> bool:
        return not node.has_children()

    @staticmethod
    def get_layers(node: Node[E]) -> list[list[Node[E]]]:
        if node is None:
            return []
        elif BinaryExpressionTree.is_leaf(node):
            return [[node]]
        output = [[node]]
        children = node.children()
        while len(children) > 0:
            output.append(children)
            children_temp = []
            for c in children:
                children_temp.extend(c.children())

            children = children_temp

        return output

    def _validate(self) -> None:
        if self._expression.count("(") != self._expression.count(")"):
            raise ValueError("Invalid Expression!")

    def _peek(self, i: typing.Optional[int] = 1) -> str:
        return self._expression[self._index + i]

    def _get_left(self) -> Node[str]:
        if self._peek(0) == "(":
            self._index += 1
            return self._get_root(")")
        lst = []
        while True:
            try:
                val = self._expression[self._index]
                if val in Operators:
                    break
                lst.append(val)
                self._index += 1
            except IndexError:
                break

        return Node("".join(lst))

    def _get_right(self, delimiter: typing.Optional[str] = None) -> Node[str]:
        idx, lst, found = self._index, [], False
        while True:
            try:
                val = self._expression[idx]
                if val == "(":
                    self._index += 1
                    return self._get_root(")")
                elif (delimiter is not None) and (val == delimiter):
                    idx += 1
                    found = True
                    break
                elif val in Operators:
                    return self._get_root(delimiter)
                lst.append(val)
                idx += 1
            except IndexError:
                break

        if (delimiter is not None) and (found is False):
            raise ValueError("Invalid Expression!")
        self._index = idx
        return Node("".join(lst))

    def _get_root(self, delimiter: typing.Optional[str] = None) -> Node[str]:
        left = self._get_left()
        root = Node(self._get_operator())
        right = self._get_right(delimiter=delimiter)
        root.left = left
        root.right = right
        return root

    def _get_operator(self) -> Operator:
        opr = self._expression[self._index]
        self._index += 1
        return typing.cast(Operator, opr)

    def parse(self) -> None:
        self._root = self._get_root()

    def __iter__(self) -> typing.Iterator[Node[E]]:
        return iter(self.breath_first_traversal())

    def __str__(self) -> str:
        if self.empty:
            return ""
        return self.visualize_top_to_bottom(self._root, self.depth)


if __name__ == "__main__":
    print("\nEnter expression, eg: (2 + 3)")
    while True:
        query = input("\n>>> ")
        if query.strip() == "q":
            break
        try:
            exp_tree = BinaryExpressionTree(query)
            exp_tree.parse()
            print(exp_tree)
        except Exception as e:
            err = "".join(traceback.format_exception(type(e), e, e.__traceback__))
            print(err)

File: BinaryExpressionTree/java/src/BinaryExpressionTree.java
import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Scanner;
import java.util.Collections;
import java.util.function.Function;

public class BinaryExpressionTree {
    private final String expression;
    private int index;
    private Node<String> root;

    private static final List<String> OPERATORS = Arrays.asList("+", "-", "/", "*");

    public BinaryExpressionTree(String expression) {
        this.expression = expression.replaceAll(" ", "");
        this.index = 0;
        this.root = null;
        this.validate();
    }

    public Node<String> getRoot() {
        return root;
    }

    public boolean isEmpty() {
        return root == null;
    }

    public static <T> boolean isLeaf(Node<T> node) {
        return !node.hasChildren();
    }

    @SuppressWarnings("all")
    public List<List<Node<String>>> getLayers() {
        if (isEmpty())
            return new ArrayList<>();
        List<List<Node<String>>> output = new ArrayList<>();
        output.add(Arrays.asList(root));
        if (isLeaf(root)) {
            return output;
        }
        List<Node<String>> children = root.children();
        while (!children.isEmpty()) {
            output.add(children);
            List<Node<String>> temp = new ArrayList<>();
            for (Node<String> child : children) {
                temp.addAll(child.children());
            }
            children = temp;
        }
        return output;
    }

    public int depth() {
        return getLayers().size();
    }

    public int height() {
        List<List<Node<String>>> layers = getLayers();
        int h = 0;
        for (List<Node<String>> list : layers) {
            h += list.size();
        }
        return h;
    }

    public int size() {
        return height();
    }

    private void validate() {
        long openB = expression.chars().filter(c -> c == '(').count();
        long closeB = expression.chars().filter(c -> c == ')').count();
        if (openB != closeB)
            throw new IllegalArgumentException("Invalid Expression!");
    }

    private String peek(int index) {
        return String.valueOf(expression.charAt(this.index + index));
    }

    private Node<String> createLeft() {
        if (peek(0).equals("(")) {
            this.index += 1;
            return createRoot(")");
        }

        List<String> list = new ArrayList<>();
        while (true) {
            try {
                String val = String.valueOf(expression.charAt(index));
                if (OPERATORS.contains(val))
                    break;
                list.add(val);
                this.index += 1;
            } catch (IndexOutOfBoundsException ignore) {
                break;
            }
        }

        return new Node<>(String.join("", list));
    }

    private Node<String> createRight(String delimiter) {
        int idx = index;
        List<String> list = new ArrayList<>();

        while (true) {
            try {
                String val = String.valueOf(expression.charAt(idx));
                if (val.equals("(")) {
                    this.index += 1;
                    return createRoot(")");
                } else if (val.equals(delimiter)) {
                    idx += 1;
                    break;
                } else if (OPERATORS.contains(val)) {
                    return createRoot(delimiter);
                }
                list.add(val);
                idx += 1;
            } catch (IndexOutOfBoundsException ignore) {
                break;
            }
        }

        this.index = idx;
        return new Node<>(String.join("", list));
    }

    private String getOperator() {
        String opr = String.valueOf(expression.charAt(index));
        this.index += 1;
        return opr;
    }

    private Node<String> createRoot(String delimiter) {
        Node<String> left = createLeft();
        Node<String> root = new Node<>(getOperator());
        Node<String> right = createRight(delimiter);
        root.setLeft(left);
        root.setRight(right);
        return root;
    }

    public void parse() {
        this.root = createRoot(null);
    }

    public void visualizeLeftToRight(Node<String> node, int level) {
        if (isEmpty() || node == null)
            return;
        if (node.getRight() != null)
            visualizeLeftToRight(node.getRight(), level + 2);
        System.out.println(" ".repeat(4).repeat(level) + "-> " + node.getElement());
        if (node.getLeft() != null)
            visualizeLeftToRight(node.getLeft(), level + 2);
    }

    public void visualizeLeftToRight() {
        visualizeLeftToRight(root, 0);
    }

    private void vis(Node<String> node, int depth, int start, List<String> list) {
        String val = node.getElement();
        int col = depth * depth;
        String nStr = " ".repeat(col) + " ".repeat(start) + val;
        list.set(node.getIndex(), nStr);
        if (node.getLeft() != null) {
            vis(node.getLeft(), depth - 1, start, list);
            if (node.getRight() != null) {
                vis(node.getRight(), depth - 1, col + start, list);
            }
        }
    }

    public String visualizeTopToBottom(Node<String> node, int depth, int start) {
        Function<List<List<Node<String>>>, Integer> createIndex = (List<List<Node<String>>> input) -> {
            List<Node<String>> flattened = new ArrayList<>();
            int counter = 0;
            for (List<Node<String>> list : input) {
                for (Node<String> n : list) {
                    n.setIndex(counter);
                    flattened.add(n);
                    counter += 1;
                }
            }
            return flattened.size();
        };

        List<List<Node<String>>> layers = getLayers();
        List<String> results = new ArrayList<>(Collections.nCopies(createIndex.apply(layers), ""));;
        vis(node, depth, start, results);
        List<String> output = new ArrayList<>();
        for (int i = 0; i < depth; i++) {
            int n = layers.get(i).size();
            List<String> list = new ArrayList<>();
            for (int j = 0; j < n; j++) {
                try {
                    String val = results.remove(0);
                    String[] split = val.split(" ");
                    int div = String.join("", list).length();
                    list.add(String.join(" ", Arrays.copyOfRange(split, div, split.length)));
                } catch (IndexOutOfBoundsException ignore) {
                    break;
                }
            }
            output.add(String.join("", list));
        }
        return String.join("\n\n", output);
    }

    public String visualizeTopToBottom() {
        if (isEmpty())
            return "";
        return visualizeTopToBottom(root, depth(), 0);
    }

    public List<Node<String>> inOrderTraversal(Node<String> node) {
        List<Node<String>> output = new ArrayList<>();
        if (isEmpty() || node == null)
            return output;
        if (node.getLeft() != null) {
            output.addAll(inOrderTraversal(node.getLeft()));
        }
        output.add(node);
        if (node.getRight() != null) {
            output.addAll(inOrderTraversal(node.getRight()));
        }
        return output;
    }

    public List<Node<String>> inOrderTraversal() {
        return inOrderTraversal(root);
    }

    public List<Node<String>> preOrderTraversal(Node<String> node) {
        List<Node<String>> output = new ArrayList<>();
        if (isEmpty() || node == null)
            return output;
        output.add(node);
        List<Node<String>> children = node.children();
        for (Node<String> child : children) {
            output.addAll(preOrderTraversal(child));
        }
        return output;
    }

    public List<Node<String>> preOrderTraversal() {
        return preOrderTraversal(root);
    }

    public List<Node<String>> postOrderTraversal(Node<String> node) {
        List<Node<String>> output = new ArrayList<>();
        if (isEmpty() || node == null)
            return output;
        if (node.getLeft() != null)
            output.addAll(postOrderTraversal(node.getLeft()));
        if (node.getRight() != null)
            output.addAll(postOrderTraversal(node.getRight()));
        output.add(node);
        return output;
    }

    public List<Node<String>> postOrderTraversal() {
        return postOrderTraversal(root);
    }

    public List<Node<String>> breathFirstTraversal() {
        List<List<Node<String>>> layers = getLayers();
        List<Node<String>> output = new ArrayList<>();
        for (List<Node<String>> list : layers) {
            output.addAll(list);
        }
        return output;
    }

    public static void main(String[] args) {
        System.out.println("\nEnter expression, eg: (2 + 3)");
        Scanner scanner = new Scanner(System.in);
        while (true) {
            System.out.print("\n>>> ");
            String query = scanner.nextLine();
            if (query.strip().equalsIgnoreCase("q"))
                break;
            try {
                BinaryExpressionTree expTree = new BinaryExpressionTree(query);
                expTree.parse();
                System.out.println();
                System.out.println(expTree.visualizeTopToBottom());
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public static class Node<E> {
        private final E element;
        private Node<E> parent;
        private Node<E> left;
        private Node<E> right;
        private int index;

        private Node(E element) {
            this.element = element;
            this.parent = null;
            this.left = null;
            this.right = null;
            this.index = -1;  // only computed when generating tree visualization
        }

        public E getElement() {
            return element;
        }

        public int getIndex() {
            return index;
        }

        public void setIndex(int index) {
            this.index = index;
        }

        public Node<E> getParent() {
            return parent;
        }

        public void setParent(Node<E> parent) {
            if (this.parent != null)
                throw new IllegalArgumentException("Parent is not empty!");
            this.parent = parent;
        }

        public Node<E> getLeft() {
            return left;
        }

        public void setLeft(Node<E> left) {
            if (this.left != null)
                throw new IllegalArgumentException("Left is not empty!");
            this.left = left;
        }

        public Node<E> getRight() {
            return right;
        }

        public void setRight(Node<E> right) {
            if (this.right != null)
                throw new IllegalArgumentException("Right is not empty!");
            if (this.left == null)
                throw new IllegalArgumentException("Left is empty!");
            this.right = right;
        }

        public List<Node<E>> children() {
            List<Node<E>> list = new ArrayList<>();
            if (left != null) {
                list.add(left);
                if (right != null) {
                    list.add(right);
                }
            }
            return list;
        }

        public boolean hasChildren() {
            return !children().isEmpty();
        }

        public String toString() {
            return String.format("%s(%s)", getClass().getName(), element);
        }
    }
}
